src/source-pcap.c

```diff
@@ -88,6 +103,8 @@ typedef struct PcapThreadVars_
     ChecksumValidationMode checksum_mode;
 
     LiveDevice *livedev;
+
+    PcapStats64 last_stats64;
```

Comment:
Simply moved the stats to the end instead of having them in the middle.

Food for thought:
A better set-up might be something like that where we'd try to stuff all the fields used in PcapCallbackLoop() into the same cache line? Assumption typical cache line of 64 bytes for Intel/AMD. Having said that, this will most likely  not make a huge difference.
```
typedef struct PcapThreadVars_
{
    /* 1ST CACHE LINE - data used in PcapCallbackLoop() */
    /* thread specific handle */
    pcap_t *pcap_handle;

    time_t last_stats_dump;

    LiveDevice *livedev;
    ThreadVars *tv;
    TmSlot *slot;

    /* counters */
    uint64_t bytes;
    uint64_t pkts;
 
    /* data link type for the thread */
    int datalink;
    
    ChecksumValidationMode checksum_mode;

    /* 2ND CACHE LINE */
    PcapStats64 last_stats64;
    uint16_t capture_kernel_packets;
    uint16_t capture_kernel_drops;
    uint16_t capture_kernel_ifdrops;

    /* ptr to string from config */
    const char *bpf_filter;

    /* thread specific bpf - pcap_setfilter() makes a copy into pcap_t */
    struct bpf_program filter;

    /** callback result -- set if one of the thread module failed. */
    int cb_result;

    /* handle state */
    unsigned char pcap_state;

    /* 3RD CACHE LINE */
    /* pcap buffer size */
    int pcap_buffer_size;
    int pcap_snaplen;
} PcapThreadVars;

```

Reply:
If you have a setup to properly test, I'm open to reshuffling. Also, we should consider the places where we count packets. I think I count 3 here: in `livedev`, in `pkts`/`bytes` and then in the stats api through `capture_kernel_*`. And then we have of course the intermediate `last_stats64` that you're introducing (for good reason). Thats a lot of stats keeping. W/o looking into why we do all this, my gut feel is that we can reduce some of this.

Reply:
Yeah, at this time I don't have a great setup at all as I am working from home. Hence, it's included as "food for thought" comment only to redo this. It should help in theory, but that needs verification. Throughput testing might not even show any improvement as all the other stuff being done with the packet might make such a change disappear in the measurement noise. I might cachegrind this later this week which should prove that it's an improvement. Until this is verified, it's probably not worth touching, imho.

While ``last_stats64`` is used together with some of the data in other cache lines, it's only updated once a second so the impact should not be too bad not getting this perfect.

However, I realized that by putting ``last_stats64`` at the end it currently straddles a cache line which we should avoid - sizeof(PcapThreadVars) is 120 before this change. I would recommend to move some config data only used in ReceivePcapThreadInit() to the end to avoid that as a minimum. This will make ``last_stats64`` completely fit into the 2nd cache line which will be better. I.e.
```
    ...
    ChecksumValidationMode checksum_mode;

    LiveDevice *livedev;

    PcapStats64 last_stats64;

    /* 3RD CACHE LINE - data only used once during ReceivePcapThreadInit() */
    /* ptr to string from config */
    const char *bpf_filter;

    /* pcap buffer size */
    int pcap_buffer_size;
    int pcap_snaplen;
} PcapThreadVars;
```

Thoughts?

Reply:
Yeah, I saw that there's a multitude of stats counters. Was wondering why there are local counters and ``livedev`` counters as well. Did not want to touch that with a ten-foot pole to not change current behaviour. I agree with your gut feel. ;)

For what it's worth, I have a version of libpcap directly providing a 64bit stats API that I will try to get upstream. That would clean the ``last_stats64`` up again if that API was detected. However, that would be a longer term change and it needs to be accepted first as well.

Reply:
@victorjulien Sorry, running short on time to do all the things on my plate and coordinate guys at work. I plan to open another version of this pull request to move the ```PcapStats64 last_stats64``` into the 2nd cache line at least. I think then that'll be ready for merge or more changes if you agree.

I can open another pull request with the reorg of the structure to be more cache friendly  for tracking purposes and to be able to pick that up another time.

Reply:
@victorjulien I can also open another ticket to revisit all the different stats collected.

Reply:
Get rid of one ticket, gain two more? At least it's progress, eh.

Reply:
Just for tracking purposes. Opened:
- Counter usages: https://redmine.openinfosecfoundation.org/issues/3829
- Cache lines: https://redmine.openinfosecfoundation.org/issues/3830